components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarInteractor.kt

```diff
@@ -0,0 +1,36 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.toolbar
+
+import mozilla.components.concept.toolbar.Toolbar
+
+/**
+ * Connects a toolbar instance to the browser engine via use cases
+ */
+class ToolbarInteractor (
+    private val loadUrlUseCase: LoadUrlUseCase,
+    private val toolbar: Toolbar
+) {
+
+    /**
+     * Starts this interactor. Makes sure this interactor is listening
+     * to relevant UI changes and triggers the corresponding use-cases
+     * in response.
+     */
+    fun start() {
+        toolbar.setOnUrlChangeListener (::loadUrl)
```

Comment:
You really like passing functions around, right? 😝

Reply:
If you look at the Kotlin bytecode in AS and decompile it to Java again then this is quite some code just for calling loadUrl().. :)

Reply:
Hehe definitely like the FP features, yes, but I think there's a bigger question here. I started off with the use-case being an interface that we can pass around, but it requires tighter coupling i.e. the toolbar-feature would then depend on the session-feature (if that's where the session use-cases) live. This is an option too. I just went with the most basic impl to start with. Wdyt? Sounds like you prefer interfaces?

Reply:
I missed that the toolbar feature module here calls use cases of the session feature module.

The only reason for this is that it needs to (indirectly) know about the session mapping. I didn't really like this thing when I wrote it. I wonder if we could prevent that. Technically this should be a use case of the toolbar feature - and it only interacts with the engine. But of course it needs to know which session. Mh.

Reply:
So, when I went back and forth on this, here's how I reasoned about this: The separation of Interactor an UseCase only makes sense if multiple Interactors should be able to invoke on the same UseCase, otherwise this logic could live in the Interactor directly. If that's the case, where should the UseCase live? It should live in the feature it depends on (see my comment above re: package by feature vs package by layer.). Let's chat about this all.